physical/consul.go

```diff
@@ -577,3 +360,270 @@ func (c *ConsulLock) Value() (bool, string, error) {
+	}
+
+	return nil
+}
+
+func (c *ConsulBackend) checkDuration() time.Duration {
+	return lib.DurationMinusBuffer(c.checkTimeout, checkMinBuffer, checkJitterFactor)
+}
+
+func (c *ConsulBackend) RunServiceDiscovery(shutdownCh ShutdownChannel, advertiseAddr string, activeFunc activeFunction, sealedFunc sealedFunction) (err error) {
+	if err := c.setAdvertiseAddr(advertiseAddr); err != nil {
+		return err
+	}
+
+	go c.runEventDemuxer(shutdownCh, advertiseAddr, activeFunc, sealedFunc)
+
+	return nil
+}
+
+func (c *ConsulBackend) runEventDemuxer(shutdownCh ShutdownChannel, advertiseAddr string, activeFunc activeFunction, sealedFunc sealedFunction) {
+	// Fire the reconcileTimer immediately upon starting the event demuxer
+	reconcileTimer := time.NewTimer(0)
+	defer reconcileTimer.Stop()
+
+	// Schedule the first check.  Consul TTL checks are passing by
+	// default, checkTimer does not need to be run immediately.
+	checkTimer := time.NewTimer(c.checkDuration())
+	defer checkTimer.Stop()
+
+	// Use a reactor pattern to handle and dispatch events to singleton
+	// goroutine handlers for execution.  It is not acceptable to drop
+	// inbound events from Notify*().
+	//
+	// goroutines are dispatched if the demuxer can acquire a lock (via
+	// an atomic CAS incr) on the handler.  Handlers are responsible for
+	// deregistering themselves (atomic CAS decr).  Handlers and the
+	// demuxer share a lock to synchronize information at the beginning
+	// and end of a handler's life (or after a handler wakes up from
+	// sleeping during a back-off/retry).
+	var shutdown bool
+	var checkLock int64
+	var registeredServiceID string
+	var serviceRegLock int64
+shutdown:
+	for {
+		select {
+		case <-c.notifyActiveCh:
+			// Run reconcile immediately upon active state change notification
+			reconcileTimer.Reset(0)
+		case <-c.notifySealedCh:
+			// Run check timer immediately upon a seal state change notification
+			checkTimer.Reset(0)
+		case <-reconcileTimer.C:
+			// Unconditionally rearm the reconcileTimer
+			reconcileTimer.Reset(reconcileTimeout - lib.RandomStagger(reconcileTimeout/checkJitterFactor))
+
+			// Abort if service discovery is disabled or a
+			// reconcile handler is already active
+			if !c.disableRegistration && atomic.CompareAndSwapInt64(&serviceRegLock, 0, 1) {
+				// Enter handler with serviceRegLock held
+				go func() {
+					defer atomic.CompareAndSwapInt64(&serviceRegLock, 1, 0)
+					for !shutdown {
+						serviceID, err := c.reconcileConsul(registeredServiceID, activeFunc, sealedFunc)
+						if err != nil {
+							c.logger.Printf("[WARN]: consul: reconcile unable to talk with Consul backend: %v", err)
+							time.Sleep(consulRetryInterval)
+							continue
+						}
+
+						c.serviceLock.Lock()
+						defer c.serviceLock.Unlock()
+
+						registeredServiceID = serviceID
+						return
+					}
+				}()
+			}
+		case <-checkTimer.C:
+			checkTimer.Reset(c.checkDuration())
+			// Abort if service discovery is disabled or a
```

Comment:
It's improbable that the service registration would not succeed but a check wouldn't, but it is theoretically possible.  Regardless, it's more important that a check run on time and on a fixed schedule, so I broke out the locks so the two handlers could run concurrently.  The check handler and reconcile handlers both read data from ConsulBackend, but neither mutate its state.  Both routines call into Vault core via `activeFunc` or `sealedFunc`, but in both cases Core handles locking there.  A single lock could work, but since the two are not interleaved, it seemed safer to burn an extra int64 and use two locks.